### DISCUSSION TITLE
authState is always null even though I have a token

### DISCUSSION THREAD

**Discussion by snaplesec** [urql-graphql/urql#1503] [Q&A] [answered]
I have been trying to use the authExchange, but I get a jwt back from server, and it never is in localStorage, and authState is never anything but null.  I have gone through the docs on this, but nothing works for me.  I am using Vue3 in vite, and apollo server with express-jwt, and jsonwebtoken.  I checked the headers and there is never an auth header, I thought that this was working before.  Ok, changed a few things, authState contains the token, but fetch fails with error:
Uncaught (in promise) TypeError: Cannot read property '__key' of undefined
    at y (fetchSource.ts:25)
    at createRequest (request.ts:81)
    at useQuery (useQuery.ts:113)
    at setup (ViewPolicyIP.vue:30)
    at callWithErrorHandling (runtime-core.esm-bundler.js:154)
    at setupStatefulComponent (runtime-core.esm-bundler.js:6484)
    at setupComponent (runtime-core.esm-bundler.js:6445)
    at mountComponent (runtime-core.esm-bundler.js:4151)
    at processComponent (runtime-core.esm-bundler.js:4127)
    at patch (runtime-core.esm-bundler.js:3745)

**Comment by JoviDeCroock** [comment, marked as answer]
That error would imply that you're not using a `query-string` which means you're just sending an empty graphql operation 😅 could you reproduce this?

**Reply by snaplesec** [reply]
When I examine the headers there is no auth header either.  I have tried many times on the query.  It works fine if I comment out the express-jwt on the server.

**Reply by snaplesec** [reply]
Here is the query:
let uname = username.value;
const { error, data } = await useQuery({
  query:` query($uname: ID!) {
        user(uname: $uname){
              userFirstName
              userLastName
              userCustomerNumber
              userName
              userActive`,
              variables: { uname } });
if (data) {
  console.log("data: ", data.value);
}
if (error) {
  console.log("Error: ", error.value);
}

**Reply by JoviDeCroock** [reply]
I mean, it would be better to show the `authExchange` as this is where the issue might reside

**Reply by snaplesec** [reply]
authExchange({
      willAuthError: () => true,
      
      getAuth: async ({ authState }) => {
        console.log("authState: ", authState);
        if (!authState) {
          const token = localStorage.getItem('token');
          if (token) {
            console.log("token in main: ", token);
            // authState = true
            return token;
          }
          return null;
        }
        // logout();
        return null;
      },
      addAuthToOperation({ operation, authState }) {
        console.log('op: %s  auth: %s', operation, authState)
        if (!authState) {
          console.log('operation: ', operation)
          return operation;
        }

        const fetchOptions =
          typeof operation.context.fetchOptions === "function"
            ? operation.context.fetchOptions()
            : operation.context.fetchOptions || {};

        return {
          ...operation.context,
          fetchOptions: {
            ...fetchOptions,
            headers: {
              ...fetchOptions.headers,
              Authorization: `Bearer ${authState}`,
            },
          },
        };
      },
    }),

**Reply by snaplesec** [reply]
addAuthToOperation() must be provided to add the custom authState to an Operation's context, so that it may be picked up by the fetchExchange.

I also see this in the addAuthToOperation code.

**Reply by snaplesec** [reply]
I'm not sure what is going on with this.  I thought it picked up the token on login before, but I am putting it in localstorage now, and that made authState contain it now at least, but it's not putting it in the fetch I guess.

**Reply by JoviDeCroock** [reply]
I mean, that makes sense you are not providing any means of fetching a token, the authExchange is meant to
- fetch a token `getAuth` --> can be sync or initial fetch
- preflight check for validity of token `willAuthError`
- postflight check for errors `didAuthError` --> refresh --> re-retrieval --> reexecution
- adding auth to the operation `addAuthToOperation`

Following the example in the docs might be a good way to get started https://formidable.com/open-source/urql/docs/advanced/authentication/

**Reply by snaplesec** [reply]
Well do you see anything wrong with mine?  I basically used the docs.  I'm not real sure what all is going on there.

**Reply by snaplesec** [reply]
Ok, so I stripped it down to minimum, and still gives same error.
app.use(urql, {
  url: url,
  fetchOptions: () => {
    const token = localStorage.getItem('token')
    return {
      headers: { "Authorization": token ? `Bearer ${token}` : ''}
    }
  }
})

**Reply by kitten** [reply]
Your `addAuthToOperation` function is returning a context object, hence the error you're seeing due to a type mismatch

**Reply by snaplesec** [reply]
How do I fix that?  I'm not sure how, I just sort of copied the examples in the docs.

**Reply by kitten** [reply]
I think you're not following the explanations in the docs closely enough, frankly speaking 😅 that being said, if you copy `addAuthToOperation` one-to-one then it will work: https://formidable.com/open-source/urql/docs/advanced/authentication/#configuring-addauthtooperation

Note that it doesn't return a context object, i.e. `operation.context` but a new cloned operation with new context. It must return a full operation, which is something that TypeScript would usually catch for you to be fair.

Also `willAuthError` cannot always return `true`, since there you're telling the exchange that it will never succeed. If you don't quite know yet what it does it's better to leave it out entirely, See: https://formidable.com/open-source/urql/docs/advanced/authentication/#configuring-willautherror

**Reply by snaplesec** [reply]
Thank you, I will try that, and I will leave out the willAuthError for now.  I do not have a refresh setup, so probably not going to work for me at this time.
